### PR TITLE
feature/1344_truncation_options_ckan

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue #1344 
* (Unofficial) e2e tests passed:

Truncation by size:

```
cygnus-ngsi.sinks.ckan-sink.truncation.max_records = 5
cygnus-ngsi.sinks.ckan-sink.truncation.max_time = -1
cygnus-ngsi.sinks.ckan-sink.truncation.checking_time = 10
..
..
time=2016-12-21T15:36:43.901UTC | lvl=DEBUG | corr=3cb04a39-4ed1-4708-956d-4df193fb75c1 | trans=3cb04a39-4ed1-4708-956d-4df193fb75c1 | srv=sc_madrid | subsrv=/traffic | comp=cygnus-ngsi | op=insert | msg=com.telefonica.iot.cygnus.backends.ckan.CKANBackendImpl[180] : Successful insert (resource/datastore id=234052d9-c736-4172-96f5-252fe2ad604d)
...
time=2016-12-21T15:37:08.352UTC | lvl=DEBUG | corr=9ec090df-68d1-4e88-884d-4061f959c6ec | trans=9ec090df-68d1-4e88-884d-4061f959c6ec | srv=sc_madrid | subsrv=/traffic | comp=cygnus-ngsi | op=insert | msg=com.telefonica.iot.cygnus.backends.ckan.CKANBackendImpl[180] : Successful insert (resource/datastore id=234052d9-c736-4172-96f5-252fe2ad604d)
...
time=2016-12-21T15:37:08.657UTC | lvl=DEBUG | corr=9ec090df-68d1-4e88-884d-4061f959c6ec | trans=9ec090df-68d1-4e88-884d-4061f959c6ec | srv=sc_madrid | subsrv=/traffic | comp=cygnus-ngsi | op=truncateBySize | msg=com.telefonica.iot.cygnus.backends.ckan.CKANBackendImpl[543] : Records must be deleted (resId=234052d9-c736-4172-96f5-252fe2ad604d, filters={"_id":[51]})
```

Truncation by time:

```
cygnus-ngsi.sinks.ckan-sink.truncation.max_records = -1
cygnus-ngsi.sinks.ckan-sink.truncation.max_time = 120
cygnus-ngsi.sinks.ckan-sink.truncation.checking_time = 10
..
..
time=2016-12-21T15:33:20.678UTC | lvl=DEBUG | corr=2b04f967-29c7-4a3a-9465-f1a94f6adb48 | trans=2b04f967-29c7-4a3a-9465-f1a94f6adb48 | srv=sc_madrid | subsrv=/traffic | comp=cygnus-ngsi | op=insert | msg=com.telefonica.iot.cygnus.backends.ckan.CKANBackendImpl[180] : Successful insert (resource/datastore id=234052d9-c736-4172-96f5-252fe2ad604d)
...
time=2016-12-21T15:35:23.795UTC | lvl=DEBUG | corr= | trans= | srv= | subsrv= | comp=cygnus-ngsi | op=truncateCachedByTime | msg=com.telefonica.iot.cygnus.backends.ckan.CKANBackendImpl[588] : Records to be deleted, resId=234052d9-c736-4172-96f5-252fe2ad604d, filters={"_id":[36,37,38,39,40,41,42,43,44,45]}
``

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -1,3 +1,4 @@
-- [cygnus-ngsi][feature] Add truncation options based on time and size (#1317)
+- [cygnus-ngsi][feature] Add global data truncation options based on time and size (#1317)
```
Comment (frbattid) [inline CHANGES_NEXT_RELEASE]: I've changed the title for the previous entry, accordingly to this new entry one.

Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: NTC

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/ckan/CKANBackendImpl.java [hunk 2]
```diff
@@ -439,6 +440,157 @@ private boolean existsView(String resId) throws Exception {
...
+            JsonResponse res = doCKANRequest("POST", urlPath, jsonString);
+
+            // check the status
+            if (res.getStatusCode() == 200) {
+                LOGGER.debug("Successful search (resourceId=" + resId + ")");
+                return res.getJsonObject();
+            } else {
+                throw new CygnusPersistenceError("Could not search for the records (resId=" + resId
+                        + ", statusCode=" + res.getStatusCode() + ")");
+            } // if else if else
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/ckan/CKANBackendImpl.java]: Too much "if else"?

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/ckan/CKANBackendImpl.java [hunk 3]
```diff
@@ -439,6 +440,157 @@ private boolean existsView(String resId) throws Exception {
...
+            // do the CKAN request
+            JsonResponse res = doCKANRequest("POST", urlPath, jsonString);
+
+            // check the status
+            if (res.getStatusCode() == 200) {
+                LOGGER.debug("Successful deletion (resourceId=" + resId + ")");
+            } else {
+                throw new CygnusPersistenceError("Could not delete the records (resId=" + resId
+                        + ", statusCode=" + res.getStatusCode() + ")");
+            } // if else if else
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/ckan/CKANBackendImpl.java]: Too much "if else"?

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md [hunk 4]
```diff
@@ -338,6 +339,9 @@ NOTE: `curl` is a Unix command allowing for interacting with REST APIs such as t
 | batch\_retry\_intervals | no | 5000 | Comma-separated list of intervals (in miliseconds) at which the retries regarding not persisted batches will be done. First retry will be done as many miliseconds after as the first value, then the second retry will be done as many miliseconds after as second value, and so on. If the batch\_ttl is greater than the number of intervals, the last interval is repeated. |
 | backend.max\_conns | no | 500 | Maximum number of connections allowed for a Http-based HDFS backend. |
 | backend.max\_conns\_per\_route | no | 100 | Maximum number of connections per route allowed for a Http-based HDFS backend. |
+| truncation.max_records | no | -1 | Maximum number of records allowed for a resource before a size-based truncation is performed. |
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md]: You should explain what `-1` means (or maybe it is done elsewhere?)

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md [hunk 5]
```diff
@@ -434,13 +441,29 @@ Finally, it must be said this way of mapping geolocated context information into
 
 [Top](#top)
 
+####<a name="section2.3.5"></a>About truncating resources
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md]: Probably to late to change... but if I'm understanding correctly the functionality not sure if "truncatnig" is the best name for it. I mean, truncating is about inserting part of the data/record/whatever instead the whole thing. What I understand you are implementing is a kind of "windowing" functionality, either based on size or time.

Or maybe I'm not understanding the functionality :). To be discussed at skype.

Comment (frbattid) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md]: As discussed offline, this will be fixed in a separate PR.

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (frbattid) [inline CHANGES_NEXT_RELEASE]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1351/commits/733214a6c45b3629c8054f3043d452340cf06c84

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/ckan/CKANBackendImpl.java (hunk 2 above)
Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/ckan/CKANBackendImpl.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1351/commits/733214a6c45b3629c8054f3043d452340cf06c84

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/ckan/CKANBackendImpl.java (hunk 3 above)
Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/ckan/CKANBackendImpl.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1351/commits/733214a6c45b3629c8054f3043d452340cf06c84

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md (hunk 4 above)
Comment (frbattid) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1351/commits/733214a6c45b3629c8054f3043d452340cf06c84

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md (hunk 5 above)
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_ckan_sink.md]: NTC

**Comment by fgalan** [comment]
LGTM
